### ISSUE TITLE
CompositeBloq in a call graph

### ISSUE THREAD

**Issue by mpharrigan** [quantumlib/Qualtran#1073]
Right now, the call graph can contain `CompositeBloq`s as nodes. 

 - composite bloqs can be quite large and quite expensive to hash and check equality, which are operations we assume are very cheap when building and manipulating a call graph.
 - It's strange to have a composite bloq inside a call graph -- that means someone has written a decomposition that has a composite bloq within it. Usually our decompositions are made out of other, concrete bloqs. If you think of bloqs as functions and compositebloq as "anonymous functions" aka lambdas, maybe this should be a supported use case but it's still strange.
 - It makes sense to me to have a compositebloq as the root node of a call graph. This is important for `CompositeBloq` to follow the `Bloq` interface and you can imagine using this to create a union of call graphs of bloqs by wrapping them in `CompositeBloq` first. There are three ways to support this
    - status quo: put the composite bloq as the root node. Hash it hopefully a limited number of times. Cache the hash of a composite bloq
    - use a sentinel value like `None` as the root node for this special case
    - Return a networkx graph that has `n` root nodes; one for each bloq instance in the composite bloq

**Comment by anurudhp** [comment]
> It's strange to have a composite bloq inside a call graph

Are there any existing bloqs where this happens? or are you just saying in general?

**Comment by mpharrigan** [comment]
I don't think we have any right now in the library. Prior to #1074, a code path would be triggered where a CompositeBloq would be the root node of the call graph

**Comment by mpharrigan** [comment]
A very similar thing is causing the confusion in #1097 where the user expects to be able to use a `CompositeBloq` as a root node of a recursive dag. In that, we can either

 - have all the `CostKey` support `CompositeBloq`; pay the cost of caching composite bloqs 
 - only allow `CompositeBloq` at the root call; special case it and omit it from the cache

**Comment by mpharrigan** [comment]
In https://github.com/quantumlib/Qualtran/issues/1097#issuecomment-2200882072, @tanujkhattar argues for full support of `CompositeBloq` including in decompositions.

My thoughts: Pro:

 - This is "morally" right and can theoretically be well-supported with the cached hash strategy alluded to in the comment
 - I think of `CompositeBloq` analogous to anonymous functions ("lambdas"). These are broadly useful in classical programming even though we don't use them in the existing bloq library yet

Considerations:
 - It's always easier to disallow something and enable it later than vice versa
 - We need more testing of examples where there is indeed a `CompositeBloq` in the call graph
